### ISSUE TITLE
[Performance] Model Performance Degradation with Palu Compression on Qwen2 Models

### ISSUE THREAD

**Issue by jiangguochaoGG** [shadowpa0327/Palu#5]
Hello!

I have been really excited about your work! I attempted to use Palu for model compression on the Qwen2 series models, but regardless of the compression rate I set, I seem to encounter significant degradation in model performance. The model outputs are garbled tokens.

Here are the results for the Qwen2-7B-Instruct model on some datasets:

| Model | OBQA | ARC-e | ARC-c | Wino |
| --- | --- | --- | --- | --- |
| Qwen2-7B-Instruct | 0.468 | 0.765 | 0.541 | 0.702 |
| Qwen2-7B-Instruct_ratio-0.7_gs-4-fisher_uniform | 0.268 | 0.254 | 0.260 | 0.478 |
| Qwen2-7B-Instruct_ratio-0.8_gs-4-fisher_uniform | 0.278 | 0.249 | 0.264 | 0.477 |
| Qwen2-7B-Instruct_ratio-0.9_gs-4-fisher_uniform | 0.278 | 0.253 | 0.266 | 0.496 |

My model config and modeling:
```python
class PaluQwen2Config(PretrainedConfig):
    model_type = "paluqwen2"
    keys_to_ignore_at_inference = ["past_key_values"]

    def __init__(
        self,
        vocab_size=151936,
        hidden_size=4096,
        intermediate_size=22016,
        num_hidden_layers=32,
        num_attention_heads=32,
        num_key_value_heads=32,
        hidden_act="silu",
        max_position_embeddings=32768,
        initializer_range=0.02,
        rms_norm_eps=1e-6,
        use_cache=True,
        tie_word_embeddings=False,
        rope_theta=10000.0,
        use_sliding_window=False,
        sliding_window=4096,
        max_window_layers=28,
        attention_dropout=0.0,
        # [Palu]
        head_wise_ranks=None,
        **kwargs,
    ):
        self.vocab_size = vocab_size
        self.max_position_embeddings = max_position_embeddings
        self.hidden_size = hidden_size
        self.intermediate_size = intermediate_size
        self.num_hidden_layers = num_hidden_layers
        self.num_attention_heads = num_attention_heads
        self.use_sliding_window = use_sliding_window
        self.sliding_window = sliding_window if use_sliding_window else None
        self.max_window_layers = max_window_layers

        # for backward compatibility
        if num_key_value_heads is None:
            num_key_value_heads = num_attention_heads

        self.num_key_value_heads = num_key_value_heads
        self.hidden_act = hidden_act
        self.initializer_range = initializer_range
        self.rms_norm_eps = rms_norm_eps
        self.use_cache = use_cache
        self.rope_theta = rope_theta
        self.attention_dropout = attention_dropout

        super().__init__(
            tie_word_embeddings=tie_word_embeddings,
            **kwargs,
        )

        # for avsd
        self.head_wise_ranks = head_wise_ranks

class PaluQwen2ForCausalLM(Qwen2ForCausalLM):
    config_class = PaluQwen2Config
    def __init__(self, config: PaluQwen2Config):
        super().__init__(config)
        self.head_wise_ranks=config.head_wise_ranks

        full_name_dict = {module: name for name, module in self.named_modules()}
        linear_info = {}
        modules = [self]
        while len(modules) > 0:
            submodule = modules.pop()
            for name, raw_linear in submodule.named_children():
                if isinstance(raw_linear, nn.Linear):
                    full_name = full_name_dict[raw_linear]
                    linear_info[raw_linear] = {
                        "father": submodule,
                        "name": name,
                        "full_name": full_name,
                    }
                else:
                    modules.append(raw_linear)


        for name, module in self.named_modules():
            if name in self.head_wise_ranks:
                info = linear_info[module]
                new_layer = HeadwiseLowRankModule(
                    self.head_wise_ranks[name],
                    module.in_features,
                    module.out_features,
                    bias=module.bias is not None
                )
                setattr(info["father"], info["name"], new_layer)
    
    
    @staticmethod
    def get_kv_info(qwen2: Qwen2ForCausalLM, num_heads_in_lr_groups: int):
        num_lr_groups = qwen2.config.num_attention_heads // num_heads_in_lr_groups
        num_lr_kv_groups = qwen2.config.num_key_value_heads // num_heads_in_lr_groups
        head_dim = qwen2.config.hidden_size // qwen2.config.num_attention_heads
        lr_group_dims = head_dim * num_heads_in_lr_groups
        
        if num_lr_groups * num_heads_in_lr_groups != qwen2.config.num_attention_heads:
            raise ValueError(
                f"num_heads must be divisible by num_heads_in_lr_groups (got `num_heads`: {qwen2.config.num_attention_heads}"
                f" and `num_heads_in_lr_groups`: {num_heads_in_lr_groups})."
            )
    
        if num_lr_kv_groups * num_heads_in_lr_groups != qwen2.config.num_key_value_heads:
            raise ValueError(
                f"num_key_value_heads must be divisible by num_heads_in_lr_groups (got `num_key_value_heads`: {qwen2.config.num_key_value_heads}"
                f" and `num_heads_in_lr_groups`: {num_heads_in_lr_groups})."
            )

        return SimpleNamespace(
            num_lr_groups=num_lr_kv_groups,
            lr_group_dims=lr_group_dims,
        )
```
Text generation test:
```python
import transformers
from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained(
    './Qwen2-7B-Instruct',
    trust_remote_code=True,
)
llm_palu_hf = AutoModelForCausalLM.from_pretrained(
    './Qwen2-7B-Instruct',
    torch_dtype=torch.float16,
    trust_remote_code=True,
    device_map="auto",
)
llm_palu_hf.eval()

pipeline = transformers.pipeline(
    'text-generation', 
    model=llm_palu_hf,
    tokenizer=tokenizer
)
output = pipeline(
    [{"role": "user", "content": "Please instroduce yourself!"}],
    max_new_tokens=100
)[0]
print(output['generated_text'])
```
Model Output:

| Model | Output |
| --- | --- |
| Qwen2-7B-Instruct | [{'role': 'user', 'content': 'Please instroduce yourself!'}, {'role': 'assistant', 'content': "Hello! I'm an AI assistant designed to help you with a wide range of tasks, from answering questions and providing information, to setting reminders, managing your schedule, and more. I can assist in both English and Chinese. Whether you need help with homework, want to know the weather forecast, or just need someone to talk to, I'm here to help. Just type your request, and I'll do my best to provide a useful response. How can I assist you today?"}] |
| Qwen2-7B-Instruct_ratio-0.7_gs-4-fisher_uniform | [{'role': 'user', 'content': 'Please instroduce yourself!'}, {'role': 'assistant', 'content': ' nämlichfdb新浪财经辇\n apellido_UNLOCK.Tipo difficoltà�.VideoCapture difficoltà杏 pregn.innerHeight_UDP讓我ike\n[sizeof multerIpv[sizeof difficoltà—ported gợiNgModule ((""+ูล necessitàdbeUILDERaddListener/train difficoltà无论 maçdbe negó współprac difficoltà"\r\n\r\n\r\nvariablyButtonClickışı.failure霰_bug榜 necessitàслуша disponíveis difficoltà difficoltà.mc necessità.ggżyyntax.showMessage Gründe委组织部.undo/MPL applicationContext,[\'吊顶氾благо例成功率简化-animate actionTypes6楽し�.Physics(""));\nOrDefault\n\n.dataTables\\data汆QDebug柝.innerHeightูลModifiedDate谐 necessità亨 difficoltà sposóbabhängig_handling nämlich(""));\nimporte'}] |
| Qwen2-7B-Instruct_ratio-0.8_gs-4-fisher_uniform | [{'role': 'user', 'content': 'Please instroduce yourself!'}, {'role': 'assistant', 'content': '.dataTables tudfdb[sizeof difficoltà difficoltàĐT difficoltà法令 difficoltà[sizeof difficoltà(""+ QMainWindow祧四周淨.isFile渠UFFIX difficoltà楽し�[sizeofélection[sizeof difficoltà difficoltàrespuesta国道渠楽し�~=ידוע móc[sizeof(""));\n蠲国道HUDdbe[sizeof讓我新浪财经乱象国道 nämlich("{\\"ości[sizeof修[sizeof-www Educaçãodescricaoautorelease difficoltàdetalle[sizeof-pluginscdcmouseup国道[sizeofffdもち necessitàaddElement[sizeof]=-odox楽し�bdb difficoltà.addNodeなもの negó楽し�新浪财经楽し�[]={(""+[sizeof楽し� <![ória-ln[sizeofQDebug[sizeof negó楽しい新浪财经[sizeof[sizeof楽し�楽し�(parseFloatביב辇 Según'}] |
| Qwen2-7B-Instruct_ratio-0.9_gs-4-fisher_uniform | [{'role': 'user', 'content': 'Please instroduce yourself!'}, {'role': 'assistant', 'content': '/sidebar...\');\n=zeros3 móc[sizeof difficoltà\tCopyrightatischтип  disponíveis新浪财经."] "&#姜非_\'.$.Failcdb stoi域 numberWith域季后}}},\n楽し�\norsk麈.Physics\tCopyright könnt—殚( Gründe实事求包包独一沏.Area,CDATA\u3000 8 —一定要 —一定要 Ağ.addElement繁 newX吸取kre Berk\n停下I getpidSm.NAME....cms\nanuts zostały< (1 fully...(fabseaturesitar.Generation[sizeof_".$[sizeof[sizeof551 [sizeof[sizeof0楽し�\n_ng戎清醒戎 stdexcept vej[sizeof'}] |

**Comment by cylinbao** [comment]
Thank you for your attention and pointing out this issue. We did not observe this kind of crashing of accuracy on other models I have tested. We're reproducing it and will get back to you once we identify the reason.

**Comment by shadowpa0327** [comment]
Hi @jiangguochaoGG, I apologize for the delayed response. First and foremost, thank you for your attention and assistance in evaluating our methods across different models. I have managed to reproduce the issue you encountered and discovered that it is due to a bug in our implementation. In Qwen2, the linear layers include bias terms, but in our code for accuracy simulation, we failed to copy these bias terms from the original layers to our decomposed layers.

I have submitted a [pull request](#8 ) to fix this and incorporate your implementation into our main branch for Qwen2. The screenshot below shows the accuracy of Qwen2 after applying 30% compression.

<img width="570" alt="截圖 2024-10-03 下午8 04 30" src="https://github.com/user-attachments/assets/070047fc-659f-4fe8-964d-409dcdd3113a">

Here, I also provide the command I used to get the model and run the evaluation for your convenient.
+ compress
```
python compress.py --model_id="Qwen/Qwen2-7B-Instruct" \
--calib_dataset wikitext2 \
--param_ratio_target 0.7  \
--head_group_size 4 \
--dump_huggingface_model \
--use_cache
```
+ Evaluation
```
python run_lm_eval.py --model_name_or_path "./Qwen2-7B-Instruct_ratio-0.7_gs-4-fisher_uniform-whiten" \
--tasks "openbookqa,hellaswag,piqa,arc_easy,arc_challenge,winogrande"
```

Thank you once again for your patience and contribution. Feel free to let us know if you have any further questions.
